src/librustc/mir/interpret/queries.rs

```diff
@@ -57,22 +55,50 @@ impl<'tcx> TyCtxt<'tcx> {
         instance: ty::Instance<'tcx>,
         span: Option<Span>,
     ) -> ConstEvalResult<'tcx> {
-        let cid = GlobalId { instance, promoted: None };
+        self.const_eval_global_id(param_env, GlobalId { instance, promoted: None }, span)
+    }
+
+    /// Evaluate a constant.
+    pub fn const_eval_global_id(
+        self,
+        param_env: ty::ParamEnv<'tcx>,
+        cid: GlobalId<'tcx>,
+        span: Option<Span>,
+    ) -> ConstEvalResult<'tcx> {
+        // Const-eval shouldn't depend on lifetimes at all, so we can erase them, which should
+        // improve caching of queries.
+        let inputs = self.erase_regions(&param_env.and(cid));
```

Comment:
I'd argue that to make this impossible to misuse, this shouldn't return `Result<ty::Const<'tcx>, ...>` (which `ConstEvalResult<'tcx>` is presumably an alias for), but just the `val` field, in the `Ok` case.

Maybe even the query shouldn't return `ty::Const<'tcx>`, but @oli-obk might have had a good reason for that.

Reply:
The idea was that we don't need to intern the result at every use site of the query. If perf says changing this to just the value, that's fine with me, but should be its own PR

Reply:
> The idea was that we don't need to intern the result at every use site of the query.

But we have to if the type differs. Maybe we should reuse the interned result from the query if the type *doesn't* differ?

Reply:
Looking at the usages of const-eval, most actually don't need to have the `Const` interned as they immediately us it. So maybe we should leave the interning of the `Const` to the caller as most of the time it doesn't actually need to be done.

Reply:
`ty::Const` is interned, and the type is inside it (the `ty` field).

So if the query interned a `ty::Const` and we have to change the `ty` field (to put in the original, unerased, type) and re-intern, it's a bit wasteful.

This is why it would be nice if the query didn't return a `ty::Const` so you couldn't accidentally use an erased type.